models/GroupsService.php

```diff
@@ -15,130 +15,148 @@
-        return $groups;
+        return array_map(
+            fn (string $group): core_kernel_classes_Resource => $this->getModel()->getResource($group),
+            $user->getPropertyValues(self::PROPERTY_MEMBERS_URI)
+        );
     }
-    
+
     /**
-     * gets the users of a group
+     * Gets the users of a group.
      *
-     * @param string $groupUri
-     * @return array resources of users
+     * @return core_kernel_classes_Resource[] User resources
      */
-    public function getUsers($groupUri)
+    public function getUsers(string $groupUri): array
     {
-        $subjectClass = TestTakerService::singleton()->getRootClass();
-        $users = $subjectClass->searchInstances([
-            self::PROPERTY_MEMBERS_URI => $groupUri
-        ], [
-            'recursive' => true, 'like' => false
-        ]);
-        return $users;
+        return $this->getTestTakerService()->getRootClass()->searchInstances(
+            [self::PROPERTY_MEMBERS_URI => $groupUri],
+            ['recursive' => true, 'like' => false]
+        );
     }
-    
+
     /**
-     * Add a User to a Group
-     *
-     * @param string $userUri
-     * @param core_kernel_classes_Resource $group
-     * @return boolean
+     * Adds a user to a Group.
      */
-    public function addUser($userUri, core_kernel_classes_Resource $group)
+    public function addUser(string $userUri, core_kernel_classes_Resource $group): bool
     {
-        $user = new \core_kernel_classes_Resource($userUri);
-        return $user->setPropertyValue(new core_kernel_classes_Property(self::PROPERTY_MEMBERS_URI), $group);
+        return $this->getModel()->getResource($userUri)->setPropertyValue(
+            $this->getModel()->getProperty(self::PROPERTY_MEMBERS_URI),
+            $group
+        );
     }
-    
+
     /**
-     * Remove a User from a Group
+     * Removes a user from a Group.
+     */
+    public function removeUser(string $userUri, core_kernel_classes_Resource $group): bool
+    {
+        return $this->getModel()->getResource($userUri)->removePropertyValue(
+            $this->getModel()->getProperty(self::PROPERTY_MEMBERS_URI),
+            $group
+        );
+    }
+
+    /**
+     * Creates a duplicate of the given group instance into the given class,
+     * copying associations for former Test Takers and Deliveries to the new group.
      *
-     * @param string $userUri
-     * @param core_kernel_classes_Resource $group
-     * @return boolean
+     * @throws common_Exception
+     * @throws common_exception_Error
+     * @throws FileExistsException
      */
-    public function removeUser($userUri, core_kernel_classes_Resource $group)
+    public function cloneInstance(
+        core_kernel_classes_Resource $instance,
+        core_kernel_classes_Class $class = null
+    ): core_kernel_classes_Resource {
+        $newGroup = parent::cloneInstance($instance, $class);
+
+        // Test takers assigned to the group are not copied by the parent class
```

Comment:
Shall we move those comments to the method doc instead? Wdyt?

Reply:
Marking this thread as solved, please do reopen it if you think there's something missing related to this topic.